index.js

```diff
@@ -91,26 +101,39 @@ async function init() {
 		const parsedUrl = new URL(query.get('url'));
 		[, user, repository, ref, dir] = urlParserRegex.exec(parsedUrl.pathname);
 
-		console.log('Source:', {user, repository, ref, dir});
+		isGHE = parsedUrl.hostname !== 'github.com';
+		if (isGHE) {
+			domain = parsedUrl.hostname;
+			apiEndpoint = `${parsedUrl.protocol}//${parsedUrl.host}/api/v3`;
+
+			document.querySelector('#create-token').href = `${parsedUrl.protocol}//${parsedUrl.host}/settings/tokens/new?description=Download GitHub directory&scopes=repo`;
+		}
+
+		console.log('Source:', {domain, user, repository, ref, dir});
 	} catch {
 		return updateStatus();
 	}
 
+	const token = await waitForToken(domain);
+
 	if (!navigator.onLine) {
 		updateStatus('⚠ You are offline.');
 		throw new Error('You are offline');
 	}
 
 	updateStatus('Retrieving directory info…');
 
-	const {private: repoIsPrivate} = await fetchRepoInfo(`${user}/${repository}`);
+	const {private: repoIsPrivate} = await fetchRepoInfo(apiEndpoint, token, `${user}/${repository}`);
 
 	const files = await listContent.viaTreesApi({
-		user,
-		repository,
-		ref,
-		directory: decodeURIComponent(dir),
-		token: localStorage.token,
+		resource: {
+			api: apiEndpoint,
+			user,
+			repository,
+			ref,
+			directory: decodeURIComponent(dir),
+		},
```

Comment:
To keep things simple on this side, you can just pass `resource: parsedUrl.href`

I think at this point you can move most of URL parsing to the only function that needs it: `fetchRepoInfo`. `fetchRepoInfo` would be changed to also just accept just `parsedUrl.pathname`

Reply:
Again, I can pass down the URL but in any case data will still be parsed at top of the `init()` so other sections can use it.
Let me know which method you prefer.

Reply:
> so other sections can use it.

My suggestion was that "other sections" don't need it.

- `fetchRepoInfo` only needs `parsedUrl.pathname` (never the domain since it's only called on github.com)
- `fetchPublicFile` needs exactly the same info, but adjusted to drop the `/tree/` part of the URL. Example:

```js
const urlParts = pathname.split('/', 5); // Ignores the directory
urlParts.splice(3, 1); // Drop `/tree/`
const response = await fetch(`https://raw.githubusercontent.com${urlParts.join('/')}/${file.path}`)
```

In short, the URL "parsing" is limited and we don't need to carry these variables around since we're taking a URL apart and then joining it together sliiiightly differently, as shown above.

Reply:
I see, thanks for the hint. In an other hand, there are some more references in the `download` function. Not sure if parsing it again would be clean enough. WDYT?

Reply:
You may be right, I didn't see that `dir` in there.

Reply:
So we keep the current changes? This can be done in another PR.

Reply:
Yes